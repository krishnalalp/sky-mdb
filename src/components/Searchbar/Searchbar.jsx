import React, { useState, useRef, useEffect } from "react";
import { useHistory, useLocation } from "react-router-dom";
import { useDispatch } from "react-redux";
import { Button, Radio } from "antd";
import { SearchOutlined } from "@ant-design/icons";
import { multiSearch } from "../../app/reducer";
import constants from "../../helpers/constants";
import { Suggestions } from "../Suggestions/Suggestions";

import "./Searchbar.scss";

export function Search() {
  const appHistory = useHistory();
  const location = useLocation();
  const isHomePage = location.pathname === "/";
  const node = useRef();
  const inputRef = useRef();

  const dispatch = useDispatch();
  const [searchTerm, setSearchTerm] = useState("");
  const [searchOption, setSearchOption] = useState(
    constants.DEFAULT_SEARCH_OPTION
  );
  const [showSuggestions, setShowSuggestions] = useState(false);

  /**
   * To show suggestions when the entered text length is 5 or above
   */
  const handleSearch = (keyword) => {
    setSearchTerm(keyword);
    if (keyword.length > constants.MIN_SEARCH_LENGTH) {
      dispatch(multiSearch(keyword, searchOption, true));
      setShowSuggestions(true);
    } else {
      setShowSuggestions(false);
    }
  };

  /**
   * To hide the suggestions section when clicked elsewhere in web page
   */
  const handleClick = (e) => {
    if (node.current.contains(e.target)) {
      return;
    }
    setShowSuggestions(false);
  };

  /**
   * To redirect to the details page when a suggestion item is clicked
   */
  const handleSuggestionClick = (path) => {
    appHistory.push(path);
    setShowSuggestions(false);
  };

  /**
   * To initiate a search/recommended movies by redirecting to the specified URL
   */
  const initiateSearch = (option) => {
    if (searchTerm) {
      appHistory.push(`/search?query=${searchTerm}&options=${option}`);
    }
    if (!searchTerm) {
      appHistory.push("/");
    }
  };

  /**
   * To initiate a search when search button is clicked or when enter is pressed
   */
  const submitSearch = (e) => {
    e.preventDefault();
    setShowSuggestions(false);
    initiateSearch(searchOption);
  };

  /**
   * To initiate a search when the search filter changes
   */
  const onSearchOptionChange = (e) => {
    const { value } = e.target;
    setSearchOption(value);
    initiateSearch(value);
  };

  /**
   * To populate the search input field with the search term on page load
   */
  useEffect(() => {
    const keyword = location.search.split("&")[0].split("=")[1];
    const option =
      location.search.split("options=")[1] || constants.DEFAULT_SEARCH_OPTION;
    setSearchOption(option);
    if (keyword) {
      setSearchTerm(decodeURIComponent(keyword));
    }
    if (isHomePage) {
      inputRef.current.focus();
    }
  }, []);

  /**
   * To handle the clicks outside the element
   */
  useEffect(() => {
    document.addEventListener("mousedown", handleClick);
    return () => {
      document.removeEventListener("mousedown", handleClick);
    };
  }, []);

  return (
    <div className="searchbar" ref={node}>
      <form onSubmit={(e) => submitSearch(e)}>
        <div className="textbox-wrapper">
          <input
            ref={inputRef}
            className="textbox"
            aria-label="Search here"
            placeholder="Movies, TV Shows, Cast"
            value={searchTerm}
            onChange={(e) => handleSearch(e.target.value)}
          />
        </div>
        <Button
          onClick={(e) => submitSearch(e)}
          type="primary"
          shape="circle"
          className="search-button"
          icon={<SearchOutlined />}
        />
        {showSuggestions && (
          <div className="suggestions-wrapper">
            <Suggestions handleClick={handleSuggestionClick} />
          </div>
        )}
      </form>
      <Radio.Group
        options={constants.SEARCH_OPTIONS}
        onChange={onSearchOptionChange}
        value={searchOption}
        optionType="button"
        buttonStyle="solid"
      />
    </div>
  );
}
