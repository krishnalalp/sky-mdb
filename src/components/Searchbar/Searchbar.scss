.searchbar {
  display: flex;
  flex-direction: column;
  padding-left: 30px;
  width: 50%;
  position: relative;
  min-width: 230px;
  padding-top: 10px;
}

.textbox-wrapper, form {
  flex: 1;
  position: relative;
}

.textbox {
  width: 100%;
  padding: 0px 20px;
  outline: 0;
  border-radius: 40px;
  border: 1px solid #fff;
  background: #000;
  color: #fff;
  height: 44px;
}

.search-button {
  position: absolute;
  top: 11px;
  right: 1px;
  width: 44px;
  height: 42px;
  border-radius: 40px;
  border: none;
}

.suggestions-wrapper {
  position: absolute;
  top: 53px;
  left: 16px;
  width: calc(100% - 35px);
  background: #000;
  z-index: 10;
  border: 1px solid #fff;
  cursor: pointer;
}