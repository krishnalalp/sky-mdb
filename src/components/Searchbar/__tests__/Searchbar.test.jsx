import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import { useLocation } from "react-router-dom";
import { Search } from "../Searchbar";

test("SearchBar component rendered", () => {
  useLocation.mockImplementation(() => ({ pathname: "/movie/299537" }));

  const renderer = new ShallowRenderer();

  renderer.render(<Search />);
  const tree = renderer.getRenderOutput();
  expect(tree).toMatchSnapshot();
});
