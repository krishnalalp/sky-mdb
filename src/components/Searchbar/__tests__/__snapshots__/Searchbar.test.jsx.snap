// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar component rendered 1`] = `
<div
  className="searchbar"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="textbox-wrapper"
    >
      <input
        aria-label="Search here"
        className="textbox"
        onChange={[Function]}
        placeholder="Movies, TV Shows, Cast"
        value=""
      />
    </div>
    <Button
      className="search-button"
      icon={<ForwardRef(SearchOutlined) />}
      onClick={[Function]}
      shape="circle"
      type="primary"
    />
  </form>
  <Memo
    buttonStyle="solid"
    onChange={[Function]}
    optionType="button"
    options={
      Array [
        Object {
          "label": "All",
          "value": "multi",
        },
        Object {
          "label": "Movies",
          "value": "movie",
        },
        Object {
          "label": "TV Shows",
          "value": "tv",
        },
        Object {
          "label": "Cast",
          "value": "person",
        },
      ]
    }
    value="multi"
  />
</div>
`;
